### DISCUSSION TITLE
Dynamic type provider configuration.

### DISCUSSION THREAD

**Discussion by voroninp** [dotnet/fsharp#13030] [General]
**Is your feature request related to a problem? Please describe.**
Due to the nature of type providers all their static parameters must be literals. It's fine until you want to implement CI/CD pipeline and then you need to configure type providers dynamically regarding build process. Each type provider solves this issue individually. Some let you define the path to local file which will be tried as a fallback, if resource is not available at given url.

**Describe the solution you'd like**
It would be very nice if F# compiler could query existing assembly for the values of static parameters.
```FSharp
[<GetParametersFrom("TypeProviderValues.dll", ProviderId="Bar1")>]
type Foo1 = Bar<>

[<GetParametersFrom("TypeProviderValues.dll", ProviderId="Bar2")>]
type Foo2 = Bar<>
```

And then assembly `TypeProviderValues.dll` which would adhere to some known strict protocol could return values for all required and optional parameters.

For example this assembly could use environment variables, or some json files or any other sources for populating parameters.

**Comment by Happypig375** [comment]
Resource values? Literal fields?

**Comment by voroninp** [comment]
@Happypig375  Do you offer Generation of literals as a part of build process? Any known solutions libraries for this?

**Comment by Happypig375** [comment]
I mean, if the assembly `TypeProviderValues.dll` really is generated, literals can be used to ensure that value lookup will not take much time.

**Comment by voroninp** [comment]
@Happypig375  Sorry, I do not get the idea.

The assembly should be there before compilation starts. It should be either another project in the same solution, or nuget package - literally a binary artefact which exists before F# compiler starts.
So if F# cannot locate the assembly it should fail.

**Comment by NatElkins** [comment]
@voroninp I think this may be what you are looking for: https://github.com/Tarmil/FSharp.Data.LiteralProviders

**Comment by voroninp** [comment]
@NatElkins Yes, this is very useful, thanks a lot! 

Tough I think this approach is not perfect as it pollutes a bit the code with the stuff which does not strictly belong to the place where type is defined. 

And is it really a responsibility of the code to decide where literals are coming from? I'd delegate this decision to 3d party. I should not care about the source so far it can make compiler happy.
